Correctly parse errors with context or with fewer newlines

shivjm opened the pull request:
I’ve added tests for the two errors I discovered were incorrectly parsed. The rectifications required tweaking the whitespace matched by the error patterns, so I updated any failing tests as well. Everything passes for me at present. Do let me know in case anything isn’t working correctly!

Closes #21

shivjm committed:
Correctly parse errors with extra context

shivjm committed:
Correctly parse errors with fewer newlines

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -105,11 +105,17 @@ https://hledger.org/hledger.html#check."
+   (error bol "hledger" (optional ".exe") ": Error: " (file-name (optional alpha ":") (+ (not ":"))) ":" line ":\n"
+    (>= 2 (optional "\n") (one-or-more (or (seq (one-or-more digit) " ") (>= 2 " ")) "| " (zero-or-more nonl) "\n")) ; excerpt lines
+    "\n"
+    (message (one-or-more (zero-or-more nonl) (or "\n" string-end)))))
```
I think the last parenthesis should be deleted and added back to the next error block. Can you please double-check?

Tests still pass though which worries me: we miss test cases.

shivjm replied:
Sorry! I’d been hunting for this parenthesis because something looked off to me but I just couldn’t see that it was at the end of the previous block.

Your comment about the tests made me realize the error pattern I added subsumes the other one.

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -105,11 +105,17 @@ https://hledger.org/hledger.html#check."
     (message "\n" (one-or-more bol (zero-or-more nonl) "\n")))
 
+   ;; hledger 1.31+ error with LINE and context:
+   (error bol "hledger" (optional ".exe") ": Error: " (file-name (optional alpha ":") (+ (not ":"))) ":" line ":\n"
```
the other error blocks start with `(error` on a dedicated line. We should try following the same pattern everywhere if possible

shivjm replied:
Roger.

DamienCassou reviewed (changes requested):
thank you for your work.

shivjm commented:
Thanks for the review. The reason why the test passed in spite of the last pattern not being included is that it matches a subset of the new pattern I added. I’ve therefore now removed the original (narrower) pattern.

DamienCassou reviewed (changes requested):
Thank you very much. The result is good now. Can you please rewrite the git history so it makes sense again?

shivjm commented:
With pleasure! I think it should be fine now. The diff (edit: between the old and rewritten histories) is empty, and I believe the commit messages are correct.

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -105,11 +105,11 @@ https://hledger.org/hledger.html#check."
-    (one-or-more (or (seq (one-or-more digit) " ") (>= 2 " ")) "| " (zero-or-more nonl) "\n")                   ; excerpt lines
-    (message "\n" (one-or-more bol (zero-or-more nonl) "\n")))))
+   ;; hledger 1.26+ error with LINE and optional context:
+   (error bol "hledger" (optional ".exe") ": Error: " (file-name (optional alpha ":") (+ (not ":"))) ":" line ":\n"
```
please keep the newline after the initial "(error" as the other patterns do.

shivjm replied:
My apologies, I think I accidentally reverted that change. Thank you for catching it.

DamienCassou reviewed (changes requested):
Thank you again. One last little thing and we are good to go.

DamienCassou reviewed (approved):
Awesome, thank you!
